hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -730,7 +730,44 @@ def examples(self):
     def all_block_bounds(self):
         return self.shrink_target.all_block_bounds()
 
-    def pass_to_descendant(self):
+    @derived_value
+    def examples_by_label(self):
+        """An index of all examples grouped by their label, with
+        the examples stored in their normal index order."""
+
+        examples_by_label = defaultdict(list)
+        for ex in self.examples:
+            examples_by_label[ex.label].append(ex)
+        return dict(examples_by_label)
+
+    def calculate_descents(self):
+        """Returns a list of all pairs (i, j) such that
+        self.examples[i] is an ancestor of self.examples[j] and
+        they have the same label.
+        """
+        result = []
+
+        for ls in self.examples_by_label.values():
+            if len(ls) <= 1:
+                continue
+
+            for i, ex in enumerate(ls[:-1]):
```

Comment:
These binary searches could potentially be avoided if each example already knows the example index of its next non-descendant.

That should be pretty easy to pull off in `stop_example`, at the expense of an extra field in `Example`:

```python
ex.end = self.index
ex.end_index = len(self.examples)
```

Reply:
Oh, that's a good idea, thanks. It requires a bit more book keeping than that (that gives us the index in the list of all examples, not the index in the list of groupings by label) but that's easily solved.

Reply:
Ah, wait, no it's not, because the following example might have a different label. Unless I'm missing a trick I don't think this binary search is avoidable.

Reply:
Oh, I think you're right. Yeah, my idea doesn't actually work, because we're searching through `ls` rather than the full example list.

Reply:
(I was fooled by the fact that I tried it, and it appeared to work. But I think it was just accidentally taking all of `ls` and trying a bunch of extra bogus shrinks.)

Reply:
Yeah, a lot of things will "almost" work here in the sense that they won't be invalid, but one of the constraints of this method is that it is worst-case quadratic but it's _really_ nice to be able to run it when the example is large because of its high probability of producing large shrinks.